Backpack/src/main/java/net/skyscanner/backpack/calendar/presenter/BpkCalendarController.kt

```diff
@@ -27,15 +27,32 @@ abstract class BpkCalendarController {
 
   abstract val locale: Locale
 
+  var selectionType: SelectionType = SelectionType.RANGE
+
   abstract fun onRangeSelected(range: CalendarRange)
 
+  abstract fun onSingleDaySelected(day: CalendarDay)
```

Comment:
I think a better and more scalable approach would be to make the selection abstract and keep only one callback. We can keep the name `onRangeSelected` to keep it backwards compatible and avoid a breaking change. 

My idea would be something like this:
```kotlin
abstract fun onRangeSelected(selection: CalendarSelection) {
    if (selection is CalendarRange) {
       // handles range
    } else if (selection is SingleDay) {
       // handle single selection
    }
}
```

This has three main benefits:
1. Doesn't need to be breaking (we just need to keep the name)
2. We don't need to keep adding more methods if a new selections are introduced (and introducing more breaking changes)
3. When subclassing this the user will be able to just care about the mode they are using. E.g. someone only using single selection can do this:

```kotlin
override fun onRangeSelected(range: CalendarRange) {
   // handle range
}
```

whereas someone using only single date can do this:

```kotlin
override fun onRangeSelected(day: CalendarDay) {
   // handle day
}
```

Reply:
If the intent is to avoid breaking changes into current classes extending `BpkCalendarController`, this won't work as on overriding `onRangeSelected()` the signature of the method needs to be kept, i.e. you need to pass a `CalendarSelection`.

This makes sense for me but just notice that the example at point 3 won't be possible.

Reply:
You're right, it would be possible to avoid breaking changes *after* this but it will be breaking now.

Reply:
👍